velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,246 @@
-        1,
-        "split only supports only single-character pattern");
-    char pattern = patternString.data()[0];
-    SplitCharacter splitCharacter(pattern);
-    splitCharacter.apply(rows, args, nullptr, context, result);
+  // Split with a non-empty delimiter. If limit > 0, The resulting array's
+  // length will not be more than limit and the resulting array's last entry
+  // will contain all input beyond the last matched regex. If limit <= 0,
+  // delimiter will be applied as many times as possible, and the resulting
+  // array can be of any size.
+  void split(
+      const StringView input,
+      const StringView delim,
+      int32_t limit,
+      vector_size_t row,
+      exec::VectorWriter<Array<Varchar>>& resultWriter) const {
+    VELOX_USER_CHECK(!delim.empty(), "Non-empty delimiter is expected");
+    // Add new array (for the new row) to our array vector.
+    resultWriter.setOffset(row);
+    auto& arrayWriter = resultWriter.current();
+
+    // Trivial case of converting string to array with 1 element.
+    if (limit == 1) {
+      arrayWriter.add_item().setNoCopy(input);
+      resultWriter.commit();
+      return;
+    }
+
+    // We walk through our input cutting off the pieces using the delimiter and
+    // add them to the elements vector, until we reach the end of the
+    // string or the limit.
+    int32_t addedElements{0};
+    auto* re = cache_.findOrCompile(delim);
+    const size_t end = input.size();
+    const char* start = input.data();
+    const auto re2String = re2::StringPiece(start, end);
+    size_t pos = 0;
+
+    re2::StringPiece subMatches[1];
+    // Matches a regular expression against a portion of the input string,
+    // starting from 'pos' to the end of the input string. The match is not
+    // anchored, which means it can start at any position in the string. If a
+    // match is found, the matched portion of the string is stored in
+    // 'subMatches'. The '1' indicates that we are only interested in the first
+    // match found from the current position 'pos' in each iteration of the
+    // loop.
+    while (re->Match(
+        re2String, pos, end, RE2::Anchor::UNANCHORED, subMatches, 1)) {
+      const auto fullMatch = subMatches[0];
+      auto offset = fullMatch.data() - start;
+      const auto size = fullMatch.size();
+      if (offset >= end) {
+        break;
+      }
+
+      // Hit an empty match for the specific case when the delimiter has '|' as
+      // a suffix. For example, if the string is 'abc' and the delimiter is 'd|'
+      // where '|' is the suffix of the pattern, using the delimiter 'd|' to
+      // split "abc" would result in an empty size match in each iteration. For
+      // empty matches, instead of inserting an empty character into the result
+      // array, always split the character at the current 'pos' of the string
+      // and put it into the result array. Thus, the result array for the above
+      // example would be ["a", "b", "c"].
```

Comment:
I find Spark's result contains an empty tail string, could you confirm?

> spark.sql("select split('abc', 'd|')").show(false)
+------------------+
|split(abc, d|, -1)|
+------------------+
|[a, b, c, ]       |
+------------------+

Reply:
yes, my mistake, actually we do contains an empty tail string and also have test case to cover https://github.com/gayangya/velox/blob/gayangya/split_refactor/velox/functions/sparksql/tests/SplitTest.cpp#L277-L282, just correct the comment.